nixos/modules/services/networking/envoy.nix

```
@@ -0,0 +1,76 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.envoy;
+
+in
+
+{
+  options.services.envoy = {
+    enable = mkEnableOption "Envoy reverse proxy";
+
+    settings = mkOption {
+      type = (pkgs.formats.json { }).type;
+      default = { };
+      example = literalExpression ''
+        {
+          admin = {
+            access_log_path = "/dev/null";
+            address = {
+              socket_address = {
+                protocol = "TCP";
+                address = "127.0.0.1";
+                port_value = 9901;
+              };
+            };
+          };
+          static_resources = {
+            listeners = [];
+            clusters = [];
+          };
+        }
+      '';
+      description = ''
+        Specify the configuration for Envoy in Nix.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.envoy ];
+    users.groups._envoy = { };
+    users.users._envoy = {
+      description = "Envoy service user";
+      group = "_envoy";
+      isSystemUser = true;
+    };
+    systemd.services.envoy = {
+      description = "Envoy reverse proxy";
+      after = [ "network-online.target" ];
+      requires = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig =
+        let
+          format = pkgs.formats.json { };
+          conf = format.generate "envoy.json" cfg.settings;
+          validateConfig = file:
+            pkgs.runCommand "validate-envoy-conf" { } ''
+              ${pkgs.envoy}/bin/envoy --log-level error --mode validate -c "${file}"
+              ln -s "${file}" "$out"
```

```suggestion
              cp "${file}" "$out"
```
if the config is not huge a symlink doesn't save a whole lot in terms of disk space, but keeping the source derivation around for the link to point to requires extra disk space for management. with a copy instead of a link the source file derivation can be garbage collected.

---

done